### PR TITLE
HTTP/2 Transport Support

### PR DESCRIPTION
**Pull request by warunalakshitha**
HTTP/2
HTTP/2 is introduced to address some drawbacks of HTTP/1 versions with features of Binary information exchange, compressed headers,multiplexing, Server Push .

Implementation
This pull request contains the initial implementation required to support HTTP/2 for current HTTP server connector including following.

Netty upgrade to 4.1.x (Required for HTTP/2)
HTTP/2 and HTTP/2 over TLS support
HTTP/2 multiplexing

### CODE REVIEW

File: http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPServerChannelInitializer.java [hunk 1]
```diff
@@ -75,14 +86,89 @@ public void initChannel(SocketChannel ch) throws Exception {
...
+     *
+     * @param ch                    Channel
+     * @param listenerConfiguration Listener Configuration
+     */
+    private void configureHttp2Pipeline(SocketChannel ch, ListenerConfiguration listenerConfiguration) {
+        ChannelPipeline p = ch.pipeline();
+        // Add http2 upgrade decoder and upgrade handler for check http version
+        final HttpServerCodec sourceCodec = new HttpServerCodec();
+        final HttpServerUpgradeHandler.UpgradeCodecFactory upgradeCodecFactory = protocol -> {
+            if (AsciiString.contentEquals(Http2CodecUtil.HTTP_UPGRADE_PROTOCOL_NAME, protocol)) {
```
Comment (Kishanthan) [inline http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPServerChannelInitializer.java:147]: inverse this if and return early for negative case

File: http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/http2/HTTP2ResponseCallback.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    private static final Logger logger = LoggerFactory.getLogger(HTTP2ResponseCallback.class);
+
+    /**
+     * Construct a new {@link HTTP2ResponseCallback} to process HTTP2 responses
+     *
+     * @param channelHandlerContext Channel context
+     * @param streamId              Stream Id
+     */
+    public HTTP2ResponseCallback(ChannelHandlerContext channelHandlerContext, int streamId) {
+
```
Comment (Kishanthan) [inline http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/http2/HTTP2ResponseCallback.java]: Unnecessary new line. please remove it.

File: http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/http2/HTTP2ResponseCallback.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        this.ctx = channelHandlerContext;
+        this.streamId = streamId;
+    }
+
+    public void done(CarbonMessage cMsg) {
+
```
Comment (Kishanthan) [inline http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/http2/HTTP2ResponseCallback.java]: Unnecessary new line. please remove it.

File: http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/http2/HTTP2SourceHandler.java [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+     * This method handles the cleartext HTTP upgrade event. If an upgrade occurred, sends a simple response via HTTP/2
+     * on stream 1 (the stream specifically reserved for cleartext HTTP upgrade).
+     *
+     * @param ctx Channel context
+     * @param evt Event
+     * @throws Exception
+     */
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+
```
Comment (Kishanthan) [inline http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/http2/HTTP2SourceHandler.java]: Unnecessary new line. please remove it.

**Review by Kishanthan — CHANGES_REQUESTED** [review]

File: http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/http2/HTTP2ResponseCallback.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+                        if (HTTPTransportContextHolder.getInstance().getHandlerExecutor() != null) {
+                            HTTPTransportContextHolder.getInstance().getHandlerExecutor()
+                                    .executeAtSourceResponseSending(cMsg);
+                        }
+                        break;
+                    }
+                    http2SourceHandler.encoder().writeData(ctx, streamId, httpContent.content(), 0, false,
+                            ctx.newPromise());
+                }
+
```
Comment (Kishanthan) [inline http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/http2/HTTP2ResponseCallback.java]: Unnecessary new line. please remove it.

File: http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/http2/HTTP2ResponseCallback.java [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+                // HTTP2 Header Request response
```
Comment (Kishanthan) [inline http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/http2/HTTP2ResponseCallback.java]: Best practice is to place the comment at the end of else if { // comment

**Review by Kishanthan — CHANGES_REQUESTED** [review]

File: http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPServerChannelInitializer.java (hunk 1 above)
Comment (warunalakshitha) [reply http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPServerChannelInitializer.java:147]: there can be multiple upgrade protocols. So if we return then that logic will break

Comment (warunalakshitha) [reply http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/HTTPServerChannelInitializer.java:147]: Other issues fixed in https://github.com/wso2/carbon-transports/pull/172

**Comment by Kishanthan** [comment]
We can close this as the same will be addressed in #172

**Pull request closed without merging**